src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -135,40 +152,92 @@ public async Task<RazorDiagnosticsResponse> Handle(RazorDiagnosticsParams reques
                 Diagnostics = mappedDiagnostics,
                 HostDocumentVersion = documentVersion,
             };
+        }
+
+        private async Task<Diagnostic[]> FilterHTMLDiagnosticsAsync(
+            Diagnostic[] unmappedDiagnostics,
+            RazorCodeDocument codeDocument,
+            DocumentSnapshot documentSnapshot)
+        {
+            var sourceText = await documentSnapshot.GetTextAsync();
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var classifiedSpans = syntaxTree.GetClassifiedSpans();
 
-            // TODO; HTML filtering blocked on https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401
-            static bool CanDiagnosticBeFiltered(RazorLanguageKind kind, Diagnostic d)
+            var processedAttributes = new Dictionary<TextSpan, bool>();
+
+            var filteredDiagnostics = unmappedDiagnostics
+                .Where(d =>
+                    // TODO: undesired HTML Diagnostic codes
+                    // (https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401) ||
+                    !InAttributeContainingCSharp(d, sourceText, syntaxTree))
+                .ToArray();
+
+            return filteredDiagnostics;
+
+            bool InAttributeContainingCSharp(Diagnostic d, SourceText sourceText, RazorSyntaxTree syntaxTree)
             {
-                return kind == RazorLanguageKind.CSharp ?
-                    CanCSharpDiagnosticBeFiltered(d) :
-                    CanHTMLDiagnosticBeFiltered(d);
+                // Examine the _end_ of the diagnostic to see if we're at the
+                // start of an (im/ex)plicit expression. Looking at the start
+                // of the diagnostic isn't sufficient.
+                var absoluteIndex = d.Range.End.GetAbsoluteIndex(sourceText);
+                var change = new SourceChange(absoluteIndex, 0, string.Empty);
+                var owner = syntaxTree.Root.LocateOwner(change);
+
+                while (owner != null)
+                {
+                    if (owner is MarkupAttributeBlockSyntax markupAttributeNode)
+                    {
+                        if (processedAttributes.TryGetValue(markupAttributeNode.FullSpan, out var doesAttributeContainsCSharp))
+                        {
+                            return doesAttributeContainsCSharp;
+                        }
+
+                        doesAttributeContainsCSharp = CheckIfAttributeContainsCSharp(markupAttributeNode.FullSpan, classifiedSpans);
+                        processedAttributes.Add(markupAttributeNode.FullSpan, doesAttributeContainsCSharp);
```

Comment:
I personally dislike utilizing locals in local functions. Instead pass the processedAttributes dictionary in.